Giwon2004/63/scheduler

giwon2004 opened the pull request:
The `scheduler.py` is implemented in a _local_ sense, which does not interact with `artiq-proxy` in a moment, and will be implemented afterward. 

The list `expList` is implemented through `QListView`, which holds `ExperimentView` instances.
The model `ExperimentView` inside `expList` also provides a drag & drop feature.

### Examples
+ Here, I had no connection through `artiq-proxy` for now, so I simply added some dummy experiments before starting.

![image](https://github.com/snu-quiqcl/iquip/assets/66408173/0803461b-5f7a-4a53-bc79-3fcb8bfe8706)
+ Current experiment is at the top, while others are on the list.
   - Other information besides their names is displayed under/beside the names.

### Problems
+ EDIT button above is not properly working, and I expect this as a problem with `QAbstractItemDelegate`.
+ Drag & drop feature may not be compatible with the practical use of `artiq-proxy`.

This closes #63.

giwon2004 committed:
Create scheduler.py

giwon2004 committed:
Create scheduler.py

giwon2004 committed:
Create scheduler.py

giwon2004 committed:
Move scheduler to `explorer.py`

giwon2004 committed:
Move scheduler to `explorer.py`

giwon2004 committed:
Move scheduler to `explorer.py`

giwon2004 committed:
Move scheduler to `explorer.py`

giwon2004 committed:
Style change

giwon2004 committed:
Style change

giwon2004 committed:
Style change

giwon2004 committed:
Style change

giwon2004 committed:
Revert "Style change"

This reverts commit 91a532549314667ff44399155dc19dfc605ebcb9.

giwon2004 committed:
Revert "Style change"

This reverts commit 91a532549314667ff44399155dc19dfc605ebcb9.

giwon2004 committed:
Revert "Style change"

This reverts commit 91a532549314667ff44399155dc19dfc605ebcb9.

giwon2004 committed:
Revert "Style change"

This reverts commit 91a532549314667ff44399155dc19dfc605ebcb9.

giwon2004 committed:
Revert "Move scheduler to `explorer.py`"

This reverts commit f6e56c843cb7b8bd8c723082bd6deaba26139e68.

giwon2004 committed:
Revert "Move scheduler to `explorer.py`"

This reverts commit f6e56c843cb7b8bd8c723082bd6deaba26139e68.

giwon2004 committed:
Revert "Move scheduler to `explorer.py`"

This reverts commit f6e56c843cb7b8bd8c723082bd6deaba26139e68.

giwon2004 committed:
Revert "Move scheduler to `explorer.py`"

This reverts commit f6e56c843cb7b8bd8c723082bd6deaba26139e68.

giwon2004 committed:
Implement Drag & Drop

giwon2004 committed:
Implement Drag & Drop

giwon2004 committed:
Implement Drag & Drop

giwon2004 committed:
Style change

giwon2004 committed:
Style change

giwon2004 committed:
Style change

giwon2004 committed:
add comments for PR

giwon2004 committed:
add comments for PR

BECATRUE commented:
Please write more descriptions about what you did and how the GUI is shown!
The PR #43 will help you.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,298 @@
+"""Protocol module for defining common forms."""
```
First of all, all developers should follow several coding rules!

- After the module docstring, add a single empty line.
- After finishing to import modules, add a single empty line. (You got it right!)
- Between the outmost classes or functions, add two empty lines.
- After the class docstring, add a single empty line.
- Indicate a type hinting for each parameter and return value.

BECATRUE replied:
Also, you should meet the Pylint condition!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,298 @@
+from iquip.protocols import ExperimentInfo
+from iquip.apps.thread import ExperimentInfoThread    
+
+def deleteItemsOfLayout(layout):
```
Please add the docstring.

giwon2004 replied:
Is it resolved now?

BECATRUE replied:
Yes, good job!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,298 @@
+        self.runExp(ExperimentInfo("exp4", {"rid": 12, "priority": 2}))
+
+    #TODO: Below are called by the signal from artiq-proxy
+    def runExp(self, info: ExperimentInfo):
```
I think moving all methods in the frame into the app looks better.

According to our philosophy, a frame is just a container with several widgets and the main controller of frame is an app.
The `builder.py` may give you some intuition.

BECATRUE replied:
I will review again after this comment is applied.

giwon2004 replied:
Is it resolved now?

BECATRUE replied:
Perfect👍

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,298 @@
+                deleteItemsOfLayout(item.layout())
+
+class SchedulerFrame(QWidget):
+    """Frame for displaying the experiment list.
```
'the submitted experiment list' looks clear.

giwon2004 committed:
Indicate parameter types

giwon2004 committed:
Indicate parameter types

giwon2004 committed:
Changes to match pylint

giwon2004 committed:
Changes to match pylint

giwon2004 committed:
Changes to match pylint (2)

giwon2004 committed:
Changes to match pylint (2)

giwon2004 committed:
Changes to match pylint (3)

giwon2004 committed:
Changes to match pylint (3)

giwon2004 committed:
Changes to match pylint (4)

giwon2004 committed:
Changes to match pylint (4)

giwon2004 committed:
Changes to match pylint (5)

giwon2004 committed:
Changes to match pylint (5)

giwon2004 committed:
Change to match pylint (6)

giwon2004 committed:
Change to match pylint (6)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,330 @@
+from PyQt5.QtGui import QFont, QPainter
+from PyQt5.QtCore import Qt, QObject, QAbstractListModel, QModelIndex, QMimeData, QModelIndex
+from PyQt5.QtWidgets import QStyleOptionViewItem, QWidget, QLayout, QLabel, QListView, \
+                            QPushButton, QHBoxLayout, QVBoxLayout, QAbstractItemDelegate, QAction
```
It is just our convention, but please modify it as follows:
```python
from PyQt5.QtWidgets import (
  QStyleOptionViewItem, ...
  QPushButton, ...
)
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,330 @@
+        self.expData = data
+
+    def rowCount(self, parent=QModelIndex()):
+        """Overrided."""
```
It looks a typo of "Overriden"😃

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,330 @@
+        if self.expData[idx].arginfo["priority"] != self.expData[taridx].arginfo["priority"]:
+            return True
+        if idx > row:
+            self.expData = self.expData[:row] + [self.expData[idx]] + \
```
Please don't use a backslash referred to [PEP 3125](https://peps.python.org/pep-3125/).
Instead, you can use parentheses.

giwon2004 committed:
Change to match pylint (7)

giwon2004 committed:
Change to match pylint (7)

giwon2004 committed:
Change format

giwon2004 committed:
Change format

giwon2004 committed:
Change to match pylint (8)

giwon2004 committed:
Change to match pylint (8)

giwon2004 committed:
Change to match pylint (9)

giwon2004 committed:
Change to match pylint (9)

giwon2004 committed:
Change to match pylint (10)

giwon2004 committed:
Change to match pylint (10)

giwon2004 committed:
Change to match pylint (11)

giwon2004 committed:
Change to match pylint (11)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+from PyQt5.QtGui import QFont, QPainter
+from PyQt5.QtCore import Qt, QObject, QAbstractListModel, QModelIndex, QMimeData
+from PyQt5.QtWidgets import (QStyleOptionViewItem, QWidget, QLayout, QLabel, QListView,
+                            QPushButton, QHBoxLayout, QVBoxLayout, QAbstractItemDelegate, QAction)
```
Could you apply the following format? Please write the imported modules in new lines.
```python
from PyQt5.QtWidgets import (
    QCheckBox, QComboBox, QDoubleSpinBox, QHBoxLayout, QLabel, QLineEdit, QListWidget,
    QListWidgetItem, QPushButton, QVBoxLayout, QWidget
)
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+
+
+class SchedulerApp(qiwis.BaseApp):
+    """App for displaying the experiment queue.
```
Please modify "... experiment" to "... submitted experiment".

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+        """Extended."""  
+        super().__init__(name, parent=parent)
+        self.schedulerFrame = SchedulerFrame()
+
```
Don't make an empty line between statements.
I think the following comment is enough to make the code legible.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+    """App for displaying the experiment queue.
+
+    Attributes:
+        schedulerFrame: The frame that shows the experiment queue.
```
"submitted"

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+        super().__init__(name, parent=parent)
+        self.schedulerFrame = SchedulerFrame()
+
+        # Below are for testing before connecting to artiq-proxy
```
These should be removed later, so please write a TODO comment.
The format is [here](https://google.github.io/styleguide/pyguide.html#312-todo-comments).

BECATRUE replied:
Never mind about getting more TODO comments.
It is more important not to be confused later.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+
+    # TODO(giwon2004): Below are called by the signal from artiq-proxy
+    def runExp(self, info: ExperimentInfo):
+        """Sets the experiment onto 'currently running' section (expRun).
```
I think that the `expRun` is unnecessary.
The following code indicates that it is connected `expRun`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+        """Extended.
+        
+        Args:
+            info: The information of the experiment.
```
Where is the argument `info`?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+        super().__init__(parent=parent)
+        layout = QVBoxLayout(self)
+        self.name = QLabel("None")
+        self.name.setFont(QFont("Arial", 10))
```
It looks good! But, you don't have to worry about design now.
Because there are no plans for design now😅

BECATRUE replied:
In my opinion, it looks better to remove it for focusing only on development.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+        """
+        super().__init__(parent=parent)
+        layout = QVBoxLayout(self)
+        self.name = QLabel("None")
```
Please set the parent explicitly.
```python
self.nameLabel = QLabel("None", self)
```

Also, please name each widget to include its type such as `nameLabel`.
It may help us to recognize the widget at a glance.

BECATRUE replied:
Please check this review!

giwon2004 replied:
It was applied, but I forgot to resolve this conversation 😅

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+                    continue
+                self.args.append(QLabel(key + ': ' + str(info.arginfo[key])))
+                self.argslayout.addWidget(self.args[-1])
+        else:
```
When is the else statement called?

The type hint of `info` is `ExperimentInfo`, I think it is always not None.

giwon2004 replied:
This is called when every experiment is done, to reset back to `None` state after termination.

BECATRUE replied:
Oh, I see. Then, I understood that the `info` may be None.
In that case, you should modify the type hinting as follows:
```python
def changeInfo(self, info: Optional[ExperimentInfo]):
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+            for key in info.arginfo:
+                if key == "priority":
+                    continue
+                self.args.append(QLabel(key + ': ' + str(info.arginfo[key])))
```
What is your purpose to store `QLabel` in both `args` and `argslayout`?
Is `args` used anywhere?

giwon2004 replied:
It was for future work, maybe we can implement a widget by clicking on each of the arguments to edit it? But it is okay to go only with `argslayout` for now. I will edit it.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,328 @@
+                if key == "priority":
+                    continue
+                self.args.append(QLabel(key + ': ' + str(info.arginfo[key])))
+                self.argslayout.addWidget(self.args[-1])
```
This is just a suggestion to show simply!

```python
for k, v in info.arginfo.items():
  if key != "priority":
    self.argslayout.addWidget(QLabel(f"{k}: {v}", self))
```

BECATRUE reviewed (changes requested):
First, I left several reviews only about `runExp()`.
I will review more about remaining parts.

giwon2004 committed:
format fix

giwon2004 committed:
format fix

giwon2004 committed:
Change to match pylint (12)

giwon2004 committed:
Change to match pylint (12)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        """Extended."""  
+        super().__init__(name, parent=parent)
+        self.schedulerFrame = SchedulerFrame()
+        # TODO(giwon2004): Below are for testing before connecting to artiq-proxy
```
It is a minor review, but please add a period(.).
Also, apply it to `runExp()`, too.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        expView = ExperimentView(info, self.parent())
+        return expView.sizeHint()
+
+    # TODO(giwon2004): Enabling buttons when displayed using QAbstractDelegate
```
Please move this into the class docstring!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        self.setLayout(layout)
+
+    def changeInfo(self, info: ExperimentInfo):
+        """Change the information by modification.
```
"Changes"

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        layout.addWidget(self.nameLabel)
+        self.setLayout(layout)
+
+    def changeInfo(self, info: ExperimentInfo):
```
I think that `updateInfo()` looks more appropriate.
Also, please write more detailed docstring!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        self.runExp(ExperimentInfo("exp4", {"rid": 12, "priority": 2}))
+
+    # TODO(giwon2004): Below are called by the signal from artiq-proxy
+    def runExp(self, info: ExperimentInfo):
```
It was discussed before, but 'exp' has too many different meanings, such as exponential, experiment, explorer, etc.
So please use the full word, i.e., `runExperiment()`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        """
+        super().__init__(parent=parent)
+        layout = QHBoxLayout(self)
+        self.argslayout = QHBoxLayout()
```
How about naming it `argsLayout`?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        for key, value in info.arginfo:
+            self.argslayout.addWidget(QLabel(f"{key}: {value}", self))
+        layout.addLayout(self.argslayout, 5)
+        self.editBtn = QPushButton("EDIT")
```
`editButton`!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+
+
+class RunningExperimentView(QWidget):
+    """Widget for displaying the information the experiment, especially for the one running.
```
Did you intend 'information of the experiment'?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        super().__init__(parent=parent)
+        layout = QVBoxLayout(self)
+        self.nameLabel = QLabel("None", self)
+        self.argslayout = QHBoxLayout()
```
Please add the parent as `self`.

giwon2004 replied:
If I add `self.argslayout = QHBoxLayout(self)` then it gives the error
```QLayout: Attempting to add QLayout "" to RunningExperimentView "", which already has a layout```.

Is it okay to move on with no changes?

BECATRUE replied:
Oh, now I realize that I didn't check it. Sorry😥

Also, I think you're right! Please remain it as it is.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        self.nameLabel = QLabel("None", self)
+        self.argslayout = QHBoxLayout()
+        layout.addWidget(self.nameLabel)
+        self.setLayout(layout)
```
This line is unnecessary! Please remove it.

BECATRUE replied:
And please follow the format of [builder.py](https://github.com/snu-quiqcl/iquip/blob/develop/iquip/apps/builder.py#L188).
```python
# widgets
...
# layout
...
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        self.setLayout(layout)
+
+class ExperimentModel(QAbstractListModel):
+    """Model for managing the data in the experiment list.
```
submitted or queued!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        data: The list of ExperimentView widget.
+    """
+
+    def __init__(self, data: list, parent=None):
```
Please add the docstring.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+
+    def __init__(self, data: list, parent=None):
+        super().__init__(parent)
+        self.expData = data
```
Please rename it to `experimentData` and update the attributes part of the class docstring.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        return self.expData[index.row()]
+
+    def supportedDropActions(self):
+        """Overrided."""
```
Overridden!

BECATRUE reviewed (changes requested):
I left several reviews!

@kangz12345 I think `ExperimentModel` class is too abstract yet, so It's hard to make reviews..

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        data: The list of ExperimentView widget.
+    """
+
+    def __init__(self, data: list, parent=None):
```
If you want to restrict the type of the list elements, you can use `typing.List`.
For example:
```python
from typing import List
def f(data: List[ExperimentView]):
    ...
```

kangz12345 replied:
Moreover, for the arguments with default values, you can write it as follows:
```python
def f(parent: Optional[QObject] = None):
    ...
```
Please add the type annotations for all the other arguments with default values as well.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        """Overrided."""
+        return self.expData[index.row()]
+
+    def supportedDropActions(self):
```
You can annotate the return type using `->` at the end of the function definition:
```python
def f() -> int:
    return 1
```
Please add the return type annotations for all the functions/methods that returns something other than `None`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        """Fetches the index of the selected element.
+        
+        Args:
+            index: The ModelIndex instance containing 
+                   row, column, and etc. values of the selected element.
```
Well, this is not a big deal at all, but our convention is to make only 2 more spaces before the overflowed lines. For example,
```python
"""...
Args:
    index: The ...
      row, column, ...
"""
```

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        mime.setText(str(index[0].row()))
+        return mime
+
+    def dropMimeData(self, # pylint: disable=too-many-arguments
+        mimedata: QMimeData,
+        action: QAction,
+        row: int,
+        column: int,
+        parentIndex: QModelIndex):
```
If there is a return value, breaking the last line would be better:
```python
def f(
    a: A,
    b: B,
) -> C:  # pylint: ...
    ...
```
Moreover, please write the pylint comment after the definition (after the colon).
Note that there are 2 spaces before `#`.

kangz12345 replied:
In addition, if a function/method returns something, it should describe the return value in its docstring. Please read [Google Python style guide](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods).

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+class ExperimentDelegate(QAbstractItemDelegate):
+    """Delegate for displaying the layout of each data in the experiment list.
+    
+    Attributes:
+        data: The list of ExperimentView widget.
```
Where did you define this?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+import qiwis
+from iquip.protocols import ExperimentInfo
+
+def delete_items_of_layout(layout: QLayout):
```
"_of_layout" seems redundant since you already named the argument informative enough.

kangz12345 replied:
To emphasize that this is an _auxiliary_ (helper) function, you can start its name with an underscore to denote that this is not a part of the public API of this module.
This is NOT mandatory!

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+    Args:
+        layout: The layout whose elements are intended to be deleted.
+    """
+    if layout is not None:
```
If you expect the case where `layout` is `None`, use type hint `layout: Optional[QLayout]`. Otherwise, this `if` statement is unnecessary.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+from iquip.protocols import ExperimentInfo
+
+def delete_items_of_layout(layout: QLayout):
+    """Auxiliary function for deleting components in the layout.
```
Does this actually _delete_ the components? Did you search for how we can delete the Qt objects properly? (because I always forget about that...)

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+    """
+
+    def __init__(self, parent: Optional[QWidget] = None):
+        """Extended. """
```
Typo: a space after `.`

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,307 @@
+        layout.addWidget(QLabel("Queued experiments:"))
+        layout.addWidget(self.expList)
+        self.setLayout(layout)
+
```
2 blank lines between class definitions.

kangz12345 commented:
> @kangz12345 I think `ExperimentModel` class is too abstract yet, so It's hard to make reviews..

I agree, please implement unit tests for it after this PR (or after any pending jobs).

kangz12345 commented:
This PR will not be reviewed by _member_ peers since it is too long to review.

giwon2004 committed:
change syntax for optional values

giwon2004 committed:
change syntax for optional values

giwon2004 committed:
Change to match pylint (13)

giwon2004 committed:
Change to match pylint (13)

giwon2004 committed:
Changes to match pylint (14)

giwon2004 committed:
Changes to match pylint (14)

giwon2004 committed:
Delete unnecessary pass

giwon2004 committed:
Delete unnecessary pass

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        """
+        self.schedulerFrame.model.experimentData.append(info)
+        self.schedulerFrame.model.experimentData.sort(key = lambda x: x.arginfo["priority"],
+                                               reverse = True)
```
Please align the starting positions of the two lines.
And remove the spaces before and after an equal sign.

```python
self.schedulerFrame.model.experimentData.sort(key=lambda x: x.arginfo["priority"],
                                              reverse=True)
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+            if widget is not None:
+                widget.setParent(None)
+                layout.removeWidget(widget)
+                widget.deleteLater()
```
👍

kangz12345 replied:
👍

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+
+    def updateInfo(self, info: Optional[ExperimentInfo] = None):
+        """Updates the information by modification.
+        This updates expRun when new experiment starts to run.
```
Please add an emtpy line after the first line of the docstring and apply this to other methods, too.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+                           self.experimentData[row:idx] + self.experimentData[idx+1:])
+        elif idx < row:
+            self.experimentData = (self.experimentData[:idx] + self.experimentData[idx+1:row] +
+                           [self.experimentData[idx]] + self.experimentData[row:])
```
Indentation!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        mime.setText(str(index[0].row()))
+        return mime
+
+    def dropMimeData(self,  # pylint: disable=too-many-arguments
```
It is just a question!
Can be only experiments with the same priority interchanged each other?

giwon2004 replied:
Yes, for now.

It is mentioned here at 2.iv in the first section: [Discussion#62](https://github.com/snu-quiqcl/iquip/discussions/62#discussion-5377377)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        idx = int(mimedata.text())
+        if action == Qt.IgnoreAction:
+            return True
+        taridx = row if row < self.rowCount() else -1
```
How about `tarIdx`?

kangz12345 replied:
"tar" is not quite a _standard_ abbreviation. Please use the full word "target". (it is target, right? 😄)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        if action == Qt.IgnoreAction:
+            return True
+        taridx = row if row < self.rowCount() else -1
+        row = row if row >= 0 else self.rowCount()
```
Redefining the existing variable doesn't look good.
How about `tarRow = row ...`?

BECATRUE replied:
Also, `row` and `taridx` looks too similar, then can you unite the two into one?

giwon2004 replied:
`taridx` was originally introduced because of "too-long" argument in `pylint`.
After that I found an error when I did not postprocess the `row` value.

I guess both can be handled with one variable. I will try to resolve this.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        Args:
+            info: The experiment information.
+        """
+        self.schedulerFrame.model.experimentData.append(info)
```
Isn't the type of `experimentData` `List[ExperimentView]`?
Thus, should be the parameter of `append()` an `ExperimentView` instance, not `ExperimentInfo`?

BECATRUE replied:
I don't know about `QAbstractListModel` well..
Due to `ExperimentDelegate`, does it work well?

giwon2004 replied:
Sorry for the confusion, `experimentData` holds the list of `ExperimentInfo`, which is rendered with `ExperimentView` (instantiated in live) in `ExperimentDelegate` as you said.

BECATRUE reviewed (changes requested):
I left several reviews.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        if (self.experimentData[idx].arginfo["priority"]
+            != self.experimentData[taridx].arginfo["priority"]):
+            return True
+        if idx > row:
+            self.experimentData = (self.experimentData[:row] + [self.experimentData[idx]] +
+                           self.experimentData[row:idx] + self.experimentData[idx+1:])
+        elif idx < row:
+            self.experimentData = (self.experimentData[:idx] + self.experimentData[idx+1:row] +
+                           [self.experimentData[idx]] + self.experimentData[row:])
```
It seems that you want to pick the element at `idx` and insert it at `row`.
How about implement it like this:
```python
if idx != row:
    item = self.experimentData.pop(idx)
    self.experimentData.insert(item, (row - 1) if idx < row else row)
```

giwon2004 replied:
Wow, I think it is very simple and straightforward. I will change it that way.

I will try my best on writing codes more neatly. Thank you!

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        return True
+
+    def flags(self, index: QModelIndex) -> int:
+        """Overrided."""
```
Overridden.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+    def updateInfo(self, info: Optional[ExperimentInfo] = None):
+        """Updates the information by modification.
+        This updates expRun when new experiment starts to run.
+        If info is None, it means there is no experiments running.
```
Please write this line in the argument docstring for `info` since it describes the API specification of `info`.
In addition, "None if there is no experiments running." sounds more natural. Make docstring (especially for API specification) for users.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+
+    def updateInfo(self, info: Optional[ExperimentInfo] = None):
+        """Updates the information by modification.
+        This updates expRun when new experiment starts to run.
```
Since `expRun` is not in this class, please clearly show where it is, e.g., `SchedulerFrame.expRun`. By the way, does this update the scheduler frame? I don't see the connection.

giwon2004 replied:
`updateInfo` in `RunningExperimentView` modifies its value without changing the id of `SchedulerFrame.expRun`. It is more like the experiment existing from the beginning changing its information repeatedly.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        return self.expInfo
+
+    def edit(self):
+        """Showing frame for edition of experiment information.
```
Please use 3rd person singular verbs for the docstring summaries.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+            self.argsLayout.addWidget(QLabel(f"{key}: {value}", self))
+
+    def data(self) -> ExperimentInfo:
+        """Data transfer for displaying in ExperimentDelegate."""
```
It is common to start with "Returns ..." for such simple _getter_ methods.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+    """
+
+    def __init__(self, data: List[ExperimentView], parent: Optional[QWidget] = None):
+        """Overridden."""
```
Since it takes more argument than its superclass' method and you called it, it is not _overridden_ but _extended_.
Please read snu-quiqcl/qiwis#44.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+        """Overridden."""
+        return len(self.experimentData)
+
+    def data(self, index: QModelIndex,
+        role: Optional[Qt.ItemDataRole] = Qt.DisplayRole
```
Break line before `index` as well.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+            row: The target row that is to be changed with the experiment in mimedata.
+            column: The target column, which is set to zero as it is a QListView.
+            parentIndex: The ModelIndex instance containing
+                         row, column, and etc. values of the target element.
```
Indentation.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,327 @@
+                         row, column, and etc. values of the target element.
+
+        Returns:
+            True value.
```
You can use [`typing.Literal`](https://docs.python.org/3/library/typing.html#typing.Literal) in this case, even if it is introduced in Python 3.8.
We are going to upgrade the Python version as #54, so we can use it soon!

giwon2004 replied:
Hmm, this gives an error in Pylint 3.7 test as you mentioned it is for 3.8+.
Is it okay to move on?

giwon2004 committed:
Update docstring with code optimization

giwon2004 committed:
Update docstring with code optimization

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,333 @@
+        super().__init__(parent)
+        self.experimentData = data
+
+    def rowCount(self, parent: Optional[QModelIndex] = QModelIndex()) -> int:
```
You can use pylint disable comment here since it overrides the parent's method.
Don't mind that the line becomes too long, because it is for technical marking using a comment.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,333 @@
+
+    def data(self,
+        index: QModelIndex,
+        role: Optional[Qt.ItemDataRole] = Qt.DisplayRole
```
Here as well.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,333 @@
+        mimedata: QMimeData,
+        action: QAction,
+        row: int,
+        column: int,
+        parentIndex: QModelIndex
```
Why are they here? Is this method also overridden?
If so, please modify the docstring and add the pylint disable comments.

giwon2004 replied:
Then do I need to delete the docstring?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,333 @@
+        expView.render(painter)
+        painter.restore()
+
+    def sizeHint(self, option: QStyleOptionViewItem, index: QModelIndex) -> QSize:
```
Here as well.

giwon2004 committed:
added pylint ignore

giwon2004 committed:
added pylint ignore

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -204,10 +206,12 @@ def dropMimeData(self,  # pylint: disable=too-many-arguments
         row: int,
-        column: int,
-        parentIndex: QModelIndex
+        column: int,  # pylint: disable=unused-argument
+        parentIndex: QModelIndex  # pylint: disable=unused-argument
```
Since they are not used, I think the argument docstring for those be "Not used." unless you are going to use it later.

kangz12345 reviewed (changes requested):
Almost there!

giwon2004 committed:
modify docstring

giwon2004 committed:
modify docstring

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        """Extended."""
+        super().__init__(parent=parent)
+        # widgets
+        self.expRun = RunningExperimentView()
+        self.expList = QListView()
```
I'm sorry to point this out now, I had no time to read it thoroughly.
We name widgets using their types, e.g., ~Button, ~Label, ~Box, etc.
This helps readers notice the widget type at a glance.
Furthermore, it clearly shows that the variable is a widget.

In this case, I would suggest `runningView` and `queueView`. Of course these are just examples. Please name them with clearly readable ones.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+    def __init__(self, parent: Optional[QWidget] = None):
+        """Extended."""
+        super().__init__(parent=parent)
+        self.experimentInfo = None
```
Missing attribute docstring.

kangz12345 replied:
Missing attribute docstring.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        self.experimentInfo = None
+        # widgets
+        self.nameLabel = QLabel("None", self)
+        self.args = []
```
Missing attribute docstring.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+                    self.argsLayout.addWidget(QLabel(f"{key}: {value}", self))
+        else:
+            self.nameLabel.setText("None")
+
```
2 blank lines between class definitions.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        """
+        _dismiss_items(self.argsLayout)
+        self.experimentInfo = info
+        if info:
```
When you test if something is `None`, you should use `if info is not None:` to prevent any unexpected false-positive.
For example, for some reason `ExperimentInfo` is updated to implement the boolean evaluation of itself. Then `if info` may return `False` even if it is not `None`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+            self.nameLabel.setText("None")
+
+class ExperimentView(QWidget):
+    """Widget for displaying the information the experiment.
```
...the information _of_ the experiment.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+            info: The information of the experiment.
+        """
+        super().__init__(parent=parent)
+        self.experimentInfo = info
```
Missing attribute docstring.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        self.experimentInfo = info
+        # widgets
+        self.nameLabel = QLabel(info.name, self)
+        self.args = [QLabel(f"{key}: {value}", self) for key, value in info.arginfo.items()]
```
Missing attribute docstring.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        for key, value in info.arginfo.items():
+            self.argsLayout.addWidget(QLabel(f"{key}: {value}", self))
+
+    def data(self) -> ExperimentInfo:
+        """Returns data to ExperimentDelegate for displaying."""
+        return self.experimentInfo
```
Do you intend to follow some convention by naming it `data()`?
I think `info()` would be more intuitive since there is `updateInfo()`.
Or, it is okay to remove this method and directly use the public attribute `experimentInfo`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        experimentData: The list of ExperimentView widget.
+    """
+
+    def __init__(self, data: List[ExperimentInfo], parent: Optional[QWidget] = None):
```
The docstring says `data` is a list of `ExperimentView` widgets, but the type annotations says it is a list of `ExperimentInfo`. Which is correct?

giwon2004 replied:
`ExperimentModel` now takes a list of `ExperimentView` instances throughout the code.

So, `ExperimentView` would be the one.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        self.schedulerFrame.model.experimentData.sort(key=lambda x: x.arginfo["priority"],
+                                                      reverse=True)
+        for data in self.schedulerFrame.model.experimentData:
+            print(type(data))
```
Is this `print()` for debugging?
If so, please remove it.
If it will be updated later, please leave a TODO comment.
If it is for logging, please use `logging` module (see `qiwis` or the latest PRs of @Aijuh).

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+        for data in self.schedulerFrame.model.experimentData:
+            print(type(data))
+
+    def changeExperiment(self, idx: int, info: Optional[ExperimentInfo] = None):
```
For the public API, please avoid using abbreviations if possible.
`index` is not even a long word!

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,336 @@
+            idx: The index of to-be-changed experiment.
+            info: The experiment information.
+        """
+        if info:
```
`if info is not None:`

giwon2004 committed:
Change data type of `ExperimentModel`

Changed data type of `ExperimentModel` from `ExperimentInfo` to `ExperimentView`.

giwon2004 committed:
Change data type of `ExperimentModel`

Changed data type of `ExperimentModel` from `ExperimentInfo` to `ExperimentView`.

giwon2004 committed:
update docstring & delete `experimentView.info()`

giwon2004 committed:
update docstring & delete `experimentView.info()`

giwon2004 committed:
minor fix

giwon2004 committed:
minor fix

giwon2004 committed:
minor fix (2)

giwon2004 committed:
minor fix (2)

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+                widget.setParent(None)
+                layout.removeWidget(widget)
+                widget.deleteLater()
+                widget = None
```
It's just a question!
Why should we reassign `widget` to None?

giwon2004 replied:
Just in case, the variable name `widget` can hold something that must be deleted.

BECATRUE replied:
Please check it! @giwon2004

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+    """Frame for displaying the submitted experiment list.
+    
+    Attributes:
+        runningView: The ExperimentView widget displaying currently running experiment.
```
Please update `ExperimentView` to `RunningExperimentView`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        layout.setContentsMargins(0, 0, 0, 20)
+        layout.addWidget(QLabel("Currently running:"))
+        layout.addWidget(self.runningView)
+        layout.addWidget(QLabel("Queued experiments:"))
```
Please set the parent explicitly.
```python
layout.addWidget(QLabel("Queued experiments:", self)
```
It should be applied to other widgets.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+    def updateInfo(self, info: Optional[ExperimentInfo] = None):
+        """Updates the information by modification.
+        
+        This updates SchedulerFrame.runningView when new experiment starts to run.
```
Just a suggestion.

```
... when a new experiment ...
```

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+            self.nameLabel.setText(info.name)
+            for key, value in info.arginfo.items():
+                if key != "priority":
+                    self.argsLayout.addWidget(QLabel(f"{key}: {value}", self))
```
Please set the parent of the `QLabel` to `self`.

giwon2004 replied:
I guess it is already there...?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+
+        Args:
+            name: The name of the app.
+            parent: The QObject that contains this app."""
```
You __must not__ write the arguments docstring of its super class, i.e., `name` and `parent`.

It's susceptible for changing the docstring of `BaseApp` because once the docstring of `BaseApp` is updated, we should update all docstring of its subclasses manually.

Thus, we decided to use the `Extended` expression and unexplained arguments imply that they are in the super class.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+            parent: The QObject that contains this app."""
+        super().__init__(name, parent=parent)
+        self.schedulerFrame = SchedulerFrame()
+        # TODO(giwon2004): Below are for testing before connecting to artiq-proxy.
```
Now, it looks unnecessary!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        if exp in self.schedulerFrame.model.experimentData:
+            self.deleteExperiment(exp)
+
+    def addExperiment(self, exp: ExperimentView):
```
How about renaming `exp` to `experiment` or `experimentView`?
It looks a little ambiguious.

Is it something that Jiyong already mentioned?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+            self.schedulerFrame.runningView.updateInfo(exp.experimentInfo)
+        else:
+            self.schedulerFrame.runningView.updateInfo(None)
+        if exp in self.schedulerFrame.model.experimentData:
```
Is it right that this condition is always the same as the above _if_ condition `exp is not None`?
If true, I think you should merge two bodies.

giwon2004 replied:
No for the first experiment, it will be directly assigned to the 'Currently running' state.

Or, am I wrong?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+            self.deleteExperiment(exp)
+
+    def addExperiment(self, exp: ExperimentView):
+        """Adds the experiment to 'queued experiments' section (queueView).
```
It looks ok to remove '(queuedView)'.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        if info is not None:
+            self.schedulerFrame.model.experimentData[index].updateInfo(info)
+            self.schedulerFrame.model.experimentData.sort(key=lambda x: x.priority(),
+                                                      reverse=True)
```
Please align the indentation.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        
+        Args:
+            info: The information of the experiment.
+            parent: The widget that contains this widget.
```
Similar to the review in the constructor of `SchedulerApp`, you should remove the `parent` docstring!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        self.experimentInfo = info
+        # widgets
+        self.nameLabel = QLabel(info.name, self)
+        self.args = [QLabel(f"{key}: {value}", self) for key, value in info.arginfo.items()]
```
Why did you define `self.args`?
I think it is used only once in the below _for_ loop and it is able to be removed.

giwon2004 replied:
Then how do I separate the widgets and the layout?

BECATRUE replied:
Please check it, too!

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+    def edit(self):
+        """Shows frame for edition of experiment information.
+
+        It is called when editButton is pressed.
```
How about `... is clicked.`?

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+            self.argsLayout.addWidget(QLabel(f"{key}: {value}", self))
+
+    def priority(self) -> int:
+        """Returns data to ExperimentDelegate for displaying."""
```
How about `Returns the data ...`?

BECATRUE replied:
And it looks necessary to explain what value is returned.

kangz12345 replied:
The docstring should be changed to describe the specific return value, i.e., priority.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        # TODO(giwon2004): Display a frame for editing the values.
+        print("Button Clicked")
+
+    def __eq__(self, other) -> bool:
```
Please add the type hint of `other`.
If it can be assigned to any type, you can write it as `other: Any`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        """Overridden."""
+        if other is None:
+            return False
+        return self.experimentInfo.arginfo["rid"] == other.experimentInfo.arginfo["rid"]
```
You should handle the case where `other` is not an instance of `ExperimentInfo`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        """Extended.
+        
+        Args:
+            parent: The widget that contains this model."""
```
Please remove it.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,352 @@
+        
+        Args:
+            parent: The widget that contains this model."""
+        super().__init__(parent)
```
It is just our convention, but please write it to `super().__init__(parent=parent)`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -166,10 +176,13 @@ class ExperimentModel(QAbstractListModel):
+            parent: The widget that contains this model."""
         super().__init__(parent)
-        self.experimentData = data
+        self.experimentData = []
```
Since the initialization does not give any information of the element type, so you can annotate it like:
```python
self.experimentData: List[ExperimentView] = []
```
By the way, why is its name "...Data", not "...Views"?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -214,7 +227,7 @@ def dropMimeData(self,  # pylint: disable=too-many-arguments
-            mimedata: The QMimeData instance containing str value of the index.
+            mimedata: The QMimeData instance containing str value of the pre-selected index.
             action: The QtAction instance classifying the action 
               (for terminating the function when it is not dropped in the appropriate region)
```
Missing period.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -281,55 +292,60 @@ class SchedulerApp(qiwis.BaseApp):
 
     # TODO(giwon2004): Below are called by the signal from artiq-proxy.
-    def runExperiment(self, info: Optional[ExperimentInfo] = None):
+    def runExperiment(self, exp: Optional[ExperimentView] = None):
```
Please avoid using abbreviations for public API: `experiment`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -281,55 +292,60 @@ class SchedulerApp(qiwis.BaseApp):
-        self.schedulerFrame.model.experimentData.append(info)
-        self.schedulerFrame.model.experimentData.sort(key=lambda x: x.arginfo["priority"],
+        self.schedulerFrame.model.experimentData.append(exp)
+        self.schedulerFrame.model.experimentData.sort(key=lambda x: x.priority(),
```
FYI (not mandatory): [`operator.methodcaller()`](https://docs.python.org/3/library/operator.html#operator.methodcaller)
E.g.: `....sort(key=operator.methodcaller("priority"), reverse=True)`

kangz12345 reviewed (changes requested):
👍

giwon2004 committed:
format change

giwon2004 committed:
format change

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -151,20 +151,24 @@ def updateInfo(self, info: ExperimentInfo):
+        """Returns the priority for sorting.
+
+        Returns:
+            The "priority" value of the experiment.
```
As described in [Google Python style guide](https://google.github.io/styleguide/pyguide.html#doc-function-returns), you can omit the "Returns:" section if the summary starts with "Returns ..." and it is clear enough.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -151,20 +151,24 @@ def updateInfo(self, info: ExperimentInfo):
 
     def priority(self) -> int:
-        """Returns data to ExperimentDelegate for displaying."""
+        """Returns the priority for sorting.
```
It is often better to describe the function (method) itself, rather than its usage.
In other words, describe _what the function does_ instead of _for what the function is used_.
The function usage can be changed or added at any time, and the docstring might not be updated accordingly.
Suggestion: `"Returns the priority of the experiment."`

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -151,20 +151,24 @@ def updateInfo(self, info: ExperimentInfo):
+    def __eq__(self, other: Any) -> bool:
         """Overridden."""
-        if other is None:
+        if isinstance(other, ExperimentView):
```
I think you missed `not`.

kangz12345 replied:
Did you check this?

giwon2004 replied:
Oh I did not, I will change it right away.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -292,37 +293,33 @@ class SchedulerApp(qiwis.BaseApp):
         Args:
-            exp: The experimentView instance to be run. None if there is no experiements running.
+            experiment: The experimentView instance to be run.
+                        None if there is no experiements running.
```
Our convention makes only 2 more spaces:
```
Args:
    experiment: ...
      None if ...
```
By the way, you can say "The experiment running now. None if ..." since you already annotated the type of `experiment`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -292,37 +293,33 @@ class SchedulerApp(qiwis.BaseApp):
-        self.schedulerFrame.model.experimentData.sort(key=lambda x: x.priority(),
+        self.schedulerFrame.model.experimentViews.append(experiment)
+        self.schedulerFrame.model.experimentViews.sort(key=operator.methodcaller("priority"),
                                                       reverse=True)
```
1 more space before `reverse=True)`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -333,19 +330,19 @@ def changeExperiment(self, index: int, info: Optional[ExperimentInfo] = None):
-                                                      reverse=True)
+            self.schedulerFrame.model.experimentViews[index].updateInfo(info)
+            self.schedulerFrame.model.experimentViews.sort(key=operator.methodcaller("priority"),
+                                                          reverse=True)
```
Here as well.

giwon2004 committed:
minor docstring change

giwon2004 committed:
minor docstring change

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -151,10 +151,7 @@ def updateInfo(self, info: ExperimentInfo):
-        Returns:
-            The "priority" value of the experiment.
+        """Returns the "priority" value of the experiment.
         """
```
It does not have to be here. When there is only the docstring summary, please write it in a single line: `"""Returns ... experiment."""`

giwon2004 committed:
minor fix

giwon2004 committed:
minor fix

kangz12345 reviewed (dismissed):
LGTM!
You can merge this after @BECATRUE approve it.

BECATRUE reviewed (changes requested):
Please check the above two reviews!

giwon2004 committed:
Delete `editButton` and change model type to `ExeperimentInfo`

giwon2004 committed:
Delete `editButton` and change model type to `ExeperimentInfo`

giwon2004 committed:
minor change

giwon2004 committed:
minor change

giwon2004 committed:
minor change (2)

giwon2004 committed:
minor change (2)

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -106,11 +104,8 @@ class ExperimentView(QWidget):
-        layout: The list of ExperimentView widget.
-        args: The list of QLabel widget displaying the experiment information.
-        argsLayout: The HBoxLayout for displaying the experiment information besides its name.
+        argsLayout: The HBoxLayout for displaying the experiment information including its name.
```
If `argsLayout` is just the layout of this widget, you can use [`layout()`](https://doc.qt.io/qt-5/qwidget.html#layout).
Is there a reason you made this as a public attribute?

giwon2004 replied:
Hmm, I just followed `explorer.py`'s convention.

For now, I did not use it outside `ExperimentView` so I will make it local.

Maybe it can be used in the menu feature to detect the clicked region. I will add that again later if I need it.

kangz12345 replied:
Oh which part of `explorer.py`? I don't see any public attribute for a layout.

giwon2004 replied:
I think I saw something in my dream... Sorry 😂

kangz12345 reviewed (changes requested):
The code became way simpler and more readable!
It seems so intuitive because of:
1. The name `experimentQueue`
2. Data flow of abstracted data structure (`ExperimentInfo`) instead of widget (`ExperimentView`)

giwon2004 committed:
delete unnecessary attributes

giwon2004 committed:
delete unnecessary attributes

giwon2004 committed:
Correct minor error

giwon2004 committed:
Correct minor error

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -115,15 +109,14 @@ def __init__(self, info: ExperimentInfo, parent: Optional[QWidget] = None):
-        self.experimentInfo = info
         # widgets
-        self.nameLabel = QLabel(info.name, self)
+        nameLabel = QLabel(info.name, self)
```
Well, public attributes are not bad! We usually set the widgets as public attributes.
The main layout, however, can be accessed with `layout()` method which is defined in `QWidget`, so there is no need to set it a public attribute.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -0,0 +1,293 @@
+              (for terminating the function when it is not dropped in the appropriate region)
+            row: The target row that is to be changed with the experiment in mimedata.
+            column: Not used.
+            parentIndex: Not used.
```
The method `dropMimeData()` is just overridden and its argumetns already are in the `dropMimeData()` of the super class, so that I think the docstrings should be removed like `ExperimentDelegate.paint()`.

BECATRUE reviewed (changes requested):
This may be the final review!

Now the code became much more legible👍

giwon2004 committed:
minor change

giwon2004 committed:
minor change

giwon2004 commented:
(Rebased)

kangz12345 commented:
Did you rebase to make the GitHub actions (pylint and unittest) pass?

giwon2004 commented:
Yes, and I thought it was almost mandatory to rebase to merge. Did I do something wrong? :(

kangz12345 commented:
No it is not wrong, but since it re-writes the commit history, please ask us before rebasing next time :)
BTW did you have to force-push the branch after rebasing? Or did you rebase the remote branch directly (I'm not sure if it is possible though)?

kangz12345 commented:
> Yes, and I thought it was almost mandatory to rebase to merge.

Oh and what do you mean by "mandatory"? Are there other reasons than pylint?

giwon2004 commented:
> > Yes, and I thought it was almost mandatory to rebase to merge.
> 
> Oh and what do you mean by "mandatory"? Are there other reasons than pylint?

If it is merged, it must follow the history of `develop` right? So I thought this pull request must be only 'ahead' of the main branch.

giwon2004 commented:
> No it is not wrong, but since it re-writes the commit history, please ask us before rebasing next time :) BTW did you have to force-push the branch after rebasing? Or did you rebase the remote branch directly (I'm not sure if it is possible though)?

Okay I will, sorry.
I just used `$ git rebase develop`, though I had to delete one extra file that I have made before, which was not present in the HEAD of the branch when I rebased.

kangz12345 commented:
> If it is merged, it must follow the history of `develop` right? So I thought this pull request must be only 'ahead' of the main branch.

No it does not have to be ahead of the main branch!
We rebase a branch when the new changes in the updated main branch is required in the current branch.
It is totally fine to merge a branch to the updated main branch (one might find some merge conflicts though).
That's why it is called _merge_.

giwon2004 commented:
> No it does not have to be ahead of the main branch! We rebase a branch when the new changes in the updated main branch is required in the current branch. It is totally fine to merge a branch to the updated main branch (one might find some merge conflicts though). That's why it is called _merge_.

True, it would be _append_ if we had to do the 'merge' part by rebasing. Thank you!

kangz12345 reviewed (dismissed):
I believe that the rebase worked well..!

kangz12345 commented:
You can merge it after @BECATRUE approve this.

BECATRUE reviewed (dismissed):
LGTM!

Now, you can merge it.

kangz12345 commented:
@BECATRUE Please approve this again.

kangz12345 commented:
Why did you merge `develop` into this branch?

giwon2004 commented:
Oh, I did not mean to do this merge.

giwon2004 committed:
Revert "Merge branch 'develop' into giwon2004/63/scheduler"

This reverts commit d53cd0826e144d465a72bbbc58ba5076c0980f1c, reversing
changes made to 3619708722cb8c1e6ee9a7a4b7809af6233d351b.

BECATRUE reviewed (approved):
If you have pushed a branch to the remote repository __once__, the branch should not be rebase even in your local repository.

It makes several unintended commits.

BECATRUE commented:
Also, please fix the conflicts!

kangz12345 commented:
I think you merged the `develop` branch just before #85 is merged. So now the logger module conflicts.

Please follow the instructions:
1. Rebase this branch onto the `develop` (don't forget to pull the `develop` in your local repo!) in your local repository.
2. Resolve the conflicts in your local repository.
3. Force-push this branch.

kangz12345 commented:
@giwon2004 Did you read this?

giwon2004 commented:
> @giwon2004 Did you read this?

I am now reading this, sorry I was busy these days.
I will do it right away.

kangz12345 reviewed (approved):
LGTM
